### DISCUSSION TITLE
"use_speaker_embeddings" and "use_d_vector" options

### DISCUSSION THREAD

**Discussion by MatthewRomanishin** [coqui-ai/TTS#1210] [General Q&A]
I want to train a model with the same parameters as your YourTTS model you use for voice conversion in colab notebook. I looked at the "use_speaker_embeddings" and "use_d_vector" options. Both of these options are True in the config.json file. I calculated the embeddings for my dataset using compute_embeddings.py. I tried to run model training with "use_speaker_embeddings"=true and "use_d_vector"=true, got the error "ValueError: [!] Speaker embedding layer already initialized before d_vector settings.". I have read the [discussion ](https://github.com/coqui-ai/TTS/discussions/1171).

Do I understand correctly that the model can only be trained with one of the parameters equal to True? How did you train the model with both parameters equal to true? If I set only one parameter to true, how will it affect the result of the inference?

**Comment by vince62s** [comment]
read this https://github.com/coqui-ai/TTS/discussions/1058

**Reply by MatthewRomanishin** [reply]
Thank you @vince62s. Sorry, but I did not find the answer to the last two questions there.

**Reply by vince62s** [reply]
either you pick use_speaker_embeddings (set to true) to train embeddings with the model, OR you use d_vectors and feed your model with externally computed embeddings.

**Reply by MatthewRomanishin** [reply]
@vince62s Thanks a lot!

**Comment by MatthewRomanishin** [comment]
@vince62s Hi! I ran into a problem while training a Vits model using "d_vector_file". I got an error: "AttributeError: 'WN' object has no attribute 'cond_layer' ".
Have you encountered such a problem?
Do you have any ideas?

Thanks.

**Reply by catalinnega** [reply]
You should specify "d_vector_dim" to VitsArgs(). By default it is set to 0, thats why the "cond_layer" within the PositionalEncoder (which uses WaveNet (WN) layers) is not initialized.

**Reply by MatthewRomanishin** [reply]
Thank you very much! I solved this problem shortly before you answered.
I have one last question. When I use "d_vector_file", in my config.json "num_speakers" is set to 0. Is this normal behavior?
Just for voice conversion it is necessary that num_speakers be > 0

**Reply by catalinnega** [reply]
"num_speakers" is used to initialize speaker_embedding layers. For d-vectors, these layers are not created, d-vectors are used only for layer conditioning. When you are using use_speaker_embedding=False, the "num_speakers" setting will not affect your architecture. 
Seems that the voice_conversion() function expects both "use_speaker_embedding" and "use_d_vector_file" configs to be True when using d_vectors, I have no idea why..
To save the number of speakers in the config.json export, you have to manually add it before declaring the model as in https://github.com/coqui-ai/TTS/blob/c63bb481e95bd4a1ff978947d8e3e6c0bfb4177f/recipes/vctk/vits/train_vits.py#L76

**Reply by OmiWakode** [reply]
Hey, I  got the same error before passing d_vector_dim. But when I try to pass d_vector_dim in VitsArgs and VitsConfig, I get another [error.](https://github.com/coqui-ai/TTS/issues/1454) . Do you have any idea about this?

**Comment by vince62s** [comment]
if you use the vits recipe you should not face this error that deals with WaveNet when vits does not use WN but Hifigan.
